tests/core/test_artifact.py

```diff
@@ -337,6 +337,58 @@ def test_revise_artifact(df):
     artifact.delete(permanent=True)  # permanent deletion
 
 
+def test_create_external_schema(tsv_file):
+    species = ln.Feature(name="species", dtype="str").save()
+    split = ln.Feature(name="split", dtype="str").save()
+    schema = ln.Schema(features=[species, split]).save()
+
+    artifact = ln.Artifact(
+        tsv_file,
+        features={"species": "bird", "split": "train"},
+        schema=schema,
+        description="test",
+    ).save()
+    assert artifact.features.get_values() == {"species": "bird", "split": "train"}
+
+    artifact.delete(permanent=True)
+    schema.delete(permanent=True)
+    species.delete(permanent=True)
+    split.delete(permanent=True)
+
+
+def test_from_dataframe_external_schema(df):
+    species = ln.Feature(name="species", dtype="str").save()
+    split = ln.Feature(name="split", dtype="str").save()
+    external_schema = ln.Schema(features=[species, split]).save()
+
+    feat1 = ln.Feature(name="feat1", dtype="int").save()
+    feat2 = ln.Feature(name="feat2", dtype="int").save()
+    schema = ln.Schema(
+        features=[feat1, feat2], slots={"external": external_schema}, otype="DataFrame"
+    ).save()
+
+    artifact = ln.Artifact.from_dataframe(
+        df,
+        features={"species": "bird", "split": "train"},
+        schema=schema,
+        description="test dataframe with external features",
+    ).save()
+
+    assert artifact.features.get_values() == {"species": "bird", "split": "train"}
+
+    # Cleanup
+    artifact.delete(permanent=True)
+    schema.delete(permanent=True)
+    external_schema.delete(permanent=True)
+    for feature in [species, split, feat1, feat2]:
```

Comment:
This line should not be needed as deleting a schema cascades to deleting the schema-feature links. Are you sure it's needed?

Reply:
We can indeed get rid of it with the alternative design: https://github.com/laminlabs/lamindb/pull/3052